[REVIEW] Data buffer bugfix and Java bindings for getMapValue/map_lookup [skip ci]

kuhushukla opened the pull request:
WIP since it is based off of https://github.com/rapidsai/cudf/pull/6315 and some other changes to cmake that may not be totally necessary. 

The code to look at for preview is in `ColumnVector.java` ,  `HostColumnVectorCore.java` , `ColumnVectorJni.cpp` , `ColumnVectorTest.java` and `TableTest.java `

I added the schema or struct datatype to `getElement` (along with some cleanup on HostColumnVector and its core class) although eventually as a follow-on it would be really nice if all these apis took DataType to allow passing metadata about them like schema any where they are used.

This PR will be rebased after https://github.com/rapidsai/cudf/pull/6278 as well, so this is very much in-flight.

GPUtester commented:
Please update the changelog in order to start CI tests.

View the gpuCI docs [here](https://docs.rapids.ai/gpuci#open-pr-or-update-pr).

kuhushukla commented:
I am going to have to do something clever for cleaning up this PRs log history once ready but for now I have simply updated with java side changes and should be good to review. @mythrocks @revans2

revans2 commented on java/pom.xml:
```diff
@@ -350,7 +350,9 @@
                                     <arg value="-DLOGGING_LEVEL=${LOGGING_LEVEL}" />
                                     <arg value="-DCMAKE_CXX_FLAGS=${cxx.flags}"/>
-                                    <arg value="-DCMAKE_EXPORT_COMPILE_COMMANDS=${CMAKE_EXPORT_COMPILE_COMMANDS}"/>
+				    <arg value="-DCMAKE_EXPORT_COMPILE_COMMANDS=${CMAKE_EXPORT_COMPILE_COMMANDS}"/>
```
nit: indentation appears to be off.

revans2 commented on java/pom.xml:
```diff
@@ -350,7 +350,9 @@
-                                    <arg value="-DCMAKE_EXPORT_COMPILE_COMMANDS=${CMAKE_EXPORT_COMPILE_COMMANDS}"/>
+				    <arg value="-DCMAKE_EXPORT_COMPILE_COMMANDS=${CMAKE_EXPORT_COMPILE_COMMANDS}"/>
+				    <arg value="-DGPU_ARCHS=ALL"/>
+				    <arg value="-DCUDA_VERSION_MAJOR=10"/>
```
How is this supposed to work for cuda 11?

kuhushukla replied:
@mythrocks I have upmerged to ur change and removed this debug I was using so I think we are good here.

revans2 commented on java/src/main/java/ai/rapids/cudf/ColumnVector.java:
```diff
@@ -2601,6 +2601,16 @@ public Table extractRe(String pattern) throws CudfException {
   }
 
+
+  public ColumnVector getMapValue(Scalar key) {
```
We need javadocs for this API.

kuhushukla replied:
done.

revans2 commented on java/src/main/java/ai/rapids/cudf/HostColumnVectorCore.java:
```diff
@@ -107,15 +107,16 @@ public long getRowCount() {
       int end = offHeap.offsets.getInt((rowIndex + 1) * DType.INT32.getSizeInBytes());
       for (int j = start; j < end; j++) {
-        retList.add(children.get(0).getElement(j));
+        retList.add(children.get(0).getElement(j, null));
```
How are we supposed to handle a struct with a struct in it?

Also a lot of this just feels confusing to me.  Why do we need this at all?  Each child type should be self describing.  We shouldn't need it to get a anything.

kuhushukla replied:
This code is only for lists, could u elaborate if u have concerns about getStruct and getList working together?

To me this is how I see the code flow : 
List<struct<struct<int, long>>>

`getList` -> `getStruct()` -> `getStruct()` -> `getElement/readValue` for basic type and another `getElement/readValue` for basicType.
We could possibly consilidate some of the if else logic here but I dont think functionality is problem with the current flow.
Let me know what you think @revans2

kuhushukla replied:
In any case, I will cleanup and push a change for readability here in a second.

kuhushukla replied:
I have pushed (the latest commit) to rework some of the calls in `getElement` and associated private methods to make it easier to follow. Let me know @revans2 if this is more palatable and if there something I may have missed. Thanks a lot!

revans2 commented on java/src/main/native/CMakeLists.txt:
```diff
@@ -49,9 +49,51 @@ if(CMAKE_COMPILER_IS_GNUCXX)
+  "List of GPU architectures (semicolon-separated) to be compiled for. Pass 'ALL' if you want to compile for all supported GPU architectures. Empty string means to auto-detect the GPUs on the current system")
+
+if("${GPU_ARCHS}" STREQUAL "")
+  message("KUHU IN EVAL")
```
nit: please remove this.

kuhushukla replied:
done.

revans2 commented on java/src/main/native/src/ColumnVectorJni.cpp:
```diff
@@ -1075,6 +1076,23 @@ JNIEXPORT jlong JNICALL Java_ai_rapids_cudf_ColumnVector_stringReplace(JNIEnv *e
   CATCH_STD(env, 0);
 }
 
+
```
nit: extra line

kuhushukla commented:
I need to do a follow on here to figure out the way we are passing the schema when structs are nested. I think the design is in an  ok shape but may have issues as we get into more functionality with struct of structs inside lists etc.

revans2 commented:
The changes look fine, but I still don't understand why we need a schema when reading data ever.

kuhushukla commented:
> The changes look fine, but I still don't understand why we need a schema when reading data ever.

>
Mostly when the column is coming back from device, it is way to know what are its number of children. Maybe we can simplify this as just premeditating the number of children we expect? Let me know if you prefer exploring that option.

kuhushukla commented:
> > The changes look fine, but I still don't understand why we need a schema when reading data ever.
> 
> > 
> 
> Mostly when the column is coming back from device, it is way to know what are its number of children. Maybe we can simplify this as just premeditating the number of children we expect? Let me know if you prefer exploring that option.

Ah, may be I can use `getNumChildren` deom the device CV but it goes thru jni

kuhushukla commented:
I am adding getNumChildren to Host side and trying a protoype - will update soon.

kuhushukla commented:
While doing 
> I am adding getNumChildren to Host side and trying a protoype - will update soon.
>

I found a bug in how we data copied to host additionally for nested type which is not required. I have added the fix for it here and done a cleanup on test utilities we added recently. I have also removed the need for passng DataType while reading which is what drove the test utils cleanup. If these latest commits need to be on a different PR, I am happy to break this into two. I did it here for ease of getting all necessary changes without too much rebasing etc.

revans2 commented on java/src/main/java/ai/rapids/cudf/HostColumnVectorCore.java:
```diff
@@ -115,29 +125,260 @@ Object getElement(int rowIndex) {
+    int end = offHeap.offsets.getInt((rowIndex + 1) * DType.INT32.getSizeInBytes());
+    int size = end - start;
+    byte[] rawData = new byte[size];
+    if (size > 0) {
```
A few nits in here. First if it is null it does not matter what the size is.  Technically a null can have data in it that is just masked.  I would prefer to see the null check the first thing, or possibly outside of the private get* methods just at the top.

kuhushukla replied:
Moved `isNull` check

kuhushukla commented:
On a slightly related note, filed https://github.com/rapidsai/cudf/issues/6339 to make some of the calls on the host side just come from this interface. It seems like the code is ready to take this improvement next.

revans2 commented on java/src/main/java/ai/rapids/cudf/HostColumnVectorCore.java:
```diff
@@ -115,29 +125,261 @@ Object getElement(int rowIndex) {
       int start = offHeap.offsets.getInt(rowIndex * DType.INT32.getSizeInBytes());
```
Why is this not calling `getList` instead? I noticed that this does not deal with null handling either.

kuhushukla replied:
It can and should. Updated.
